compiler/rustc_codegen_llvm/src/builder/autodiff.rs

```diff
@@ -0,0 +1,363 @@
+use std::ffi::CString;
+use std::ptr;
+
+use libc::{c_char, c_uint};
+use rustc_ast::expand::autodiff_attrs::{AutoDiffAttrs, AutoDiffItem, DiffActivity, DiffMode};
+use rustc_codegen_ssa::ModuleCodegen;
+use rustc_codegen_ssa::back::write::ModuleConfig;
+use rustc_errors::FatalError;
+use rustc_session::config::Lto;
+use tracing::{debug, trace};
+
+use crate::declare::declare_raw_fn;
+use rustc_codegen_ssa::traits::{BaseTypeCodegenMethods, BuilderMethods, TypeMembershipCodegenMethods};
+//use rustc_data_structures::unord::UnordMap;
+//use rustc_hir::def_id::LOCAL_CRATE;
+use rustc_middle::ty::TyCtxt;
+//use rustc_middle::mir::mono;
+//use mono::{CodegenUnit, CodegenUnitNameBuilder};
+
+use crate::builder::Builder;
+use crate::back::write::{llvm_err, llvm_optimize};
+use crate::context;
+use crate::errors::LlvmError;
+use crate::llvm::AttributePlace::Function;
+use crate::llvm::{Metadata, True};
+use crate::value::Value;
+use crate::{CodegenContext, LlvmCodegenBackend, ModuleLlvm, attributes, llvm};
+
+fn get_params(fnc: &Value) -> Vec<&Value> {
+    unsafe {
+        let param_num = llvm::LLVMCountParams(fnc) as usize;
+        let mut fnc_args: Vec<&Value> = vec![];
+        fnc_args.reserve(param_num);
+        llvm::LLVMGetParams(fnc, fnc_args.as_mut_ptr());
+        fnc_args.set_len(param_num);
+        fnc_args
+    }
+}
+
+/// When differentiating `fn_to_diff`, take a `outer_fn` and generate another
+/// function with expected naming and calling conventions[^1] which will be
+/// discovered by the enzyme LLVM pass and its body populated with the differentiated
+/// `fn_to_diff`. `outer_fn` is then modified to have a call to the generated
+/// function and handle the differences between the Rust calling convention and
+/// Enzyme.
+/// [^1]: <https://enzyme.mit.edu/getting_started/CallingConvention/>
+// FIXME(ZuseZ4): `outer_fn` should include upstream safety checks to
+// cover some assumptions of enzyme/autodiff, which could lead to UB otherwise.
+fn generate_enzyme_call<'ll, 'tcx>(
+    llcx: &'ll llvm::Context,
```

Comment:
Get this value from CodegenCx where it's still needed (prefer adding methods tho if none exist yet)